src/circuit.rs

```diff
@@ -293,13 +291,16 @@ where
 {
     /// Circuit identifier associated constant.
     const CIRCUIT_ID: [u8; 32];
+
     /// Gadget implementation used to fill the composer.
     fn gadget(
         &mut self,
         composer: &mut StandardComposer<E, P>,
     ) -> Result<(), Error>;
+
     /// Compiles the circuit by using a function that returns a `Result`
     /// with the `ProverKey`, `VerifierKey` and the circuit size.
+    #[allow(clippy::type_complexity)] // NOTE: Clippy is too hash here.
```

Comment:
We can maybe move this to the `lib.rs` if we thing will pop up more.

Reply:
NB: I meant `harsh` there instead of `hash`.

It's generally not a good idea to dismiss `clippy` lints globally unless they are really broken or too strict like `pedantic`. Lints related to code complexity are important for constraining the design space and sometimes `clippy` is "wrong" in that it treats all complexity the same using a length heuristic, when that's not always what makes something complex or not. See for example other parts of the library where `FIXME` was the note instead for complexity-related lints. It's important to distinguish these two cases when we know the readability and design are not ideal, and when `clippy` is just not subtle enough.